source/cluster_manager/clusterManager.js

```diff
@@ -555,204 +566,945 @@ var ClusterManager = function (clusterName, selfId, spec) {
+            return Promise.resolve();
+        }
+
+        return new Promise((resolve, reject) => {
+            this._waitingPromises.push(()=>{
+                log.info("lock for:", msg)
+                resolve();
+            });
+        });
+    }
+
+    unlock() {
+        if (this._waitingPromises.length > 0) {
+            const next = this._waitingPromises.shift();
+            next();
         } else {
-            log.info('slave, not concerned message:', message);
+            log.debug("first lock for:", this.msg);
+            this._locked = false;
+        }
+    }
+}
+const mutex = new Mutex();
+const DEFAULT_DATA = -1; //Number.MIN_SAFE_INTEGER;
+
+var state = {
+    term: DEFAULT_DATA, // current term
+    leaderId: 0, // candidate and follower found leaderId
+    commitId: DEFAULT_DATA, //last sync data index
+    lastVoteTerm: DEFAULT_DATA, // candidate last vote for other candidate term
+    lastVoteFor: DEFAULT_DATA, // candidate last vote for other candidate.id
+    voteNum: 0, //candidate laste got vote from other candidate
+    totalNode: 0,
+    prevLogIndex: DEFAULT_DATA,// last log entry index
+    prevLogTerm: DEFAULT_DATA, // last log entry term
+    voters: new Set(), //who vote for me
+    role: "follower",
+    timestamp: (new Date()).getTime()
+}
+
+var leaderMsgHandler = null;
+
+var runAsFollower = function (topicChannel, manager) {
+    var isInstallSnapshot = false;
+    var verifyLeaderCallback = [];
+    var isVerifyLeader = false;
+    var verifyLeaderResult = false;
+    state.role = "follower";
+    state.timestamp=(new Date()).getTime();
+
+    var routerKeys = [`clusterManager.follower.${state.id}`];
+    var observer = new HeartbeatObserver(async (loseTime,lastContact)=>{
+        let begin = (new Date()).getTime();
+        await mutex.lock("follower.HeartbeatObserver");
+        try{
+            if(state.role !== "follower"){
+                log.warn('follower cycle in HeartbeatObserver',"self:", state);
+                return;
+            }
+
+            //check if leader is still alive
+            if (false && state.leaderId != 0) {
+                let count = 2;
+                while (count > 0) {
+                    count--;
+                    if (await verifyLeader()) {
+                        observer.resetTimer();
+                        return;
+                    }
+                }
+            }
+            log.fatal("Lose heart-beat from leader:",state.leaderId,"loseTime:", loseTime, 'lastContact:', lastContact, "timeout:",`${loseTime.getTime()-lastContact.getTime()}ms`);
+            await changeRole("candidate",state.term);
+        }catch(e){
+            log.fatal(util.inspect(e));
+        }finally {
+            let end = (new Date()).getTime();
+            let delay = end-begin;
+            if(delay>150){
+                log.debug(state.id, `follower.verifyLeader handler delay:`, `${delay}ms`,"leaderId:", state.leaderId, "self:", state);
```

Comment:
What does (delay > 150) means?

Reply:
150 specifies that the instruction processing time exceeds 150ms, and this log was added during development to debug the instruction processing time. Due to the presence of mutex.lock, the processing time of the previous instruction will affect the processing time of the next instruction. By using the handler delay in the log, it is possible to layer by layer identify which instruction is processing time consuming. When I tested here, I found that the maximum 500ms of installsnapshot would cause a delay in subsequent signal processing. Other instructions did not have this problem. I plan to change the installsnapshot method and use promise and timer to speed up the response speed

Reply:
> What does (delay > 150) mean?

150 specifies that the instruction processing time exceeds 150ms, and this log was added during development to debug the instruction processing time. Due to the presence of mutex.lock, the processing time of the previous instruction will affect the processing time of the next instruction. By using the handler delay in the log, it is possible to layer by layer identify which instruction is processing time consuming. When I tested here, I found that the maximum 500ms of installsnapshot would cause a delay in subsequent signal processing. Other instructions did not have this problem. I plan to change the installsnapshot method and use promise and timer to speed up the response speed

Reply:
Why does installsnapshot take 500ms? Have you investigated the long time of installsnapshot?
Since your check interval is 20ms, I think you can only get <20ms response time reduction by changing timer.
Maybe the 500ms response time is caused by other factors.

Reply:
i have change this in my new commit:https://github.com/open-webrtc-toolkit/owt-server/pull/1380/commits/7b6e7e490274cd3990749fe49454a33e23b1c715,if got reponse will setRunTimeData immediately